### PR TITLE
Anisha from pretrained

### PR DESCRIPTION
**Pull request by A9isha**
# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/maxtext_utils.py [hunk 1]
```diff
@@ -848,3 +853,61 @@ def schedule(step):
...
+  # Model definition
+  quant = quantizations.configure_quantization(config)
+  model = models.Transformer(config, mesh, quant=quant)
+  return model
+
+def create_training_tools(config, model, mesh, init_rng):
+  """Creates the optimizer, learning rate schedule, and checkpoint manager."""
+  learning_rate_schedule = create_learning_rate_schedule(config)
+  tx = optimizers.get_optimizer(config, learning_rate_schedule)
+  logger = checkpointing.setup_checkpoint_logger(config)
```
Comment (SurbhiJainUSC) [inline MaxText/maxtext_utils.py]: We can move line 877:911 in a new method say `checkpointing.create_checkpoint_manager()` to make this more cleaner.
https://github.com/AI-Hypercomputer/maxtext/blob/ff6c8aca080eb55cde3e26ccbcdba5fbe0ebe8b9/MaxText/checkpointing.py

File: MaxText/global_store.py
Comment (gagika) [inline MaxText/global_store.py:1]: Do we really need global_store?

one option is to keep all those as attributes of the model, e.g.

def from_pretrained(
  ...
) -> Transformer:
    """Load a pretrained MaxText model from checkpoint."""

    # Create all the necessary components
    mesh, init_rng = maxtext_utils.initialize_platform(config)
    model = maxtext_utils.create_model(config, mesh)
    checkpoint_manager, learning_rate_schedule, tx = maxtext_utils.create_training_tools(config, model, mesh, init_rng)

    #  Attach components as model attributes
    model.mesh = mesh
    model.init_rng = init_rng
    model.checkpoint_manager = checkpoint_manager
    model.learning_rate_schedule = learning_rate_schedule
    model.tx = tx

    return model
    
    And use as:
    
    if hasattr(model, 'mesh'):
        mesh = model.mesh
        init_rng = model.init_rng
        checkpoint_manager = model.checkpoint_manager
        learning_rate_schedule = model.learning_rate_schedule
        tx = model.tx
        
      
    Another option is to have dataclass with all those fields that from_pretrained returns

File: MaxText/__init__.py [hunk 2]
```diff
@@ -20,3 +20,51 @@
...
+    Example:
+        # Basic usage - mesh and init_rng are stored globally
+        model = from_pretrained(config)
+
+        # Disable global storage for testing
+        model = from_pretrained(config, store_global_state=False)
+    """
+    mesh, init_rng = maxtext_utils.initialize_platform(config)
+    model = maxtext_utils.create_model(config, mesh)
+    checkpoint_manager, learning_rate_schedule, tx = maxtext_utils.create_training_tools(config, model, mesh, init_rng)
```
Comment (SurbhiJainUSC) [inline MaxText/__init__.py]: We don't really need to create `checkpoint_manager`, `learning_rate_schedule` and `tx` inside `from_pretrained`. Also, there is no need to store them in global store. This can be moved to `setup_train_loop`. Example:
```
def setup_train_loop():
       model = mt(config)
       # Check if we have values in the global store
       global_store = global_store.get_global_store()

       # If the global store is initialized, we retrieve the training supporters from it.
       if global_store.is_initialized():
               mesh, init_rng = global_store.get_training_supporters()
       else:
               mesh, init_rng = maxtext_utils.initialize_platform(config)
    
       checkpoint_manager, learning_rate_schedule, tx = maxtext_utils.create_training_tools(config, model, mesh, init_rng)
```

File: MaxText/pyconfig.py [hunk 3]
```diff
@@ -21,6 +21,8 @@
 import math
 import os
 import sys
+import datetime
+from typing import Any, Union
```
Comment (SurbhiJainUSC) [inline MaxText/pyconfig.py]: I think this import is not used anywhere. Can be removed?

File: MaxText/checkpointing.py [hunk 4]
```diff
@@ -374,9 +374,7 @@ def setup_checkpoint_logger(config) -> Any | None:  # pytype: disable=attribute-
   max_logging.log("Setting up checkpoint logger...")
   if config.enable_checkpoint_cloud_logger:
     logger_name = f"goodput_{config.run_name}"
-    options = ocp.logging.CloudLoggerOptions(
-        job_name=config.run_name, logger_name=logger_name
-    )
+    options = ocp.logging.CloudLoggerOptions(job_name=config.run_name, logger_name=logger_name)
```
Comment (SurbhiJainUSC) [inline MaxText/checkpointing.py]: There is a PR in review for all the linting issues you are facing: https://github.com/AI-Hypercomputer/maxtext/pull/1844

Comment (A9isha) [reply MaxText/checkpointing.py]: Thank you so much!

File: MaxText/__init__.py (hunk 2 above)
Comment (A9isha) [reply MaxText/__init__.py]: done - thanks Surbhi!

File: MaxText/maxtext_utils.py (hunk 1 above)
Comment (A9isha) [reply MaxText/maxtext_utils.py]: thanks Surbhi - they now live in `create_training_tools()` in `maxtext_utils.py`

File: MaxText/pyconfig.py (hunk 3 above)
Comment (A9isha) [reply MaxText/pyconfig.py]: datetime I introduced in lines 512-515 to fill in a run_name if it was not provided, following xpk submission script
Any, Union are being used already

File: MaxText/__init__.py [hunk 5]
```diff
@@ -20,3 +20,36 @@
...
+Transformer = models.Transformer
+
+
+def from_pretrained(config: pyconfig.HyperParameters, **kwargs) -> Transformer:
+  """Load a pretrained MaxText model from checkpoint.
+
+  This function loads a model from a checkpoint.
+
+  Args:
+      config: Optional config object. If None, will load from checkpoint
```
Comment (gagika) [inline MaxText/__init__.py]: From the below implementation, the config can't be None

File: MaxText/maxtext_utils.py [hunk 6]
```diff
@@ -1078,3 +1082,55 @@ def get_formatted_sharding_annotations(params, mesh=None):
...
+
+def create_model(config, mesh):
+  """Instantiates and returns the model object, sharded across the mesh."""
+  # Model definition
+  quant = quantizations.configure_quantization(config)
+  model = models.Transformer(config, mesh, quant=quant)
+  return model
+
+
+def create_training_tools(config, model, mesh):
```
Comment (gagika) [inline MaxText/maxtext_utils.py]: nit: perhaps `setup_training_components`, `create_training_components` or `initialize_training_components` is more descriptive (e.g. tools seams very generic)

**Review by gagika — COMMENTED** [review]
Could you add some unit and integration tests? 
it can be a follow up PR.

File: MaxText/maxtext_utils.py (hunk 6 above)
Comment (A9isha) [reply MaxText/maxtext_utils.py]: sounds good - I have made it into setup_training_components - thanks Gagik

File: MaxText/__init__.py (hunk 5 above)
Comment (A9isha) [reply MaxText/__init__.py]: ah - yes - fixed comment now because config as a mandatory arg - thanks Gagik

File: MaxText/global_store.py
Comment (A9isha) [reply MaxText/global_store.py:1]: I have updated the train.py to use from_pretrained, and all the other trainers use train.py's function to get the mesh/model etc. I think the existing unit tests and integration tests are testing the new changes. What do you think?

File: MaxText/__init__.py (hunk 5 above)
Comment (gagika) [reply MaxText/__init__.py]: could you update the comment, it still has optional, none:

config: Optional config object. If None, will load from checkpoint

Comment (A9isha) [reply MaxText/__init__.py]: Sorry, updated now
